src/simsopt/mhd/virtual_casing.py

```diff
@@ -0,0 +1,355 @@
+        unit_normal = surf.unitnormal()
+        Bxyz = B_cartesian(vmec, nphi=nphi, ntheta=ntheta, range="full torus")
+        logger.debug(f'gamma.shape: {gamma.shape}')
+        logger.debug(f'unit_normal.shape: {unit_normal.shape}')
+        logger.debug(f'Bxyz[0].shape: {Bxyz[0].shape}')
+
+        # virtual_casing wants all input arrays to be 1D. The order is
+        # {x11, x12, ..., x1Np, x21, x22, ... , xNtNp, y11, ... , z11, ...}
+        # where Nt is toroidal (not theta!) and Np is poloidal (not phi!)
+        gamma1d = np.zeros(nphi * ntheta * 3)
+        B1d = np.zeros(nphi * ntheta * 3)
+        B3d = np.zeros((nphi, ntheta, 3))
+        for jxyz in range(3):
+            gamma1d[jxyz * nphi * ntheta: (jxyz + 1) * nphi * ntheta] = gamma[:, :, jxyz].flatten(order='C')
+            B1d[jxyz * nphi * ntheta: (jxyz + 1) * nphi * ntheta] = Bxyz[jxyz].flatten(order='C')
+            B3d[:, :, jxyz] = Bxyz[jxyz]
+
+        """
+        # Check order:
+        index = 0
+        for jxyz in range(3):
+            for jphi in range(nphi):
+                for jtheta in range(ntheta):
+                    np.testing.assert_allclose(gamma1d[index], gamma[jphi, jtheta, jxyz])
+                    np.testing.assert_allclose(B1d[index], Bxyz[jxyz][jphi, jtheta])
+                    index += 1
+        """
+
+        vcasing = vc_module.VirtualCasing()
+        vcasing.set_surface(nphi, ntheta, gamma1d)
+        vcasing.set_accuracy(digits)
+        # This next line launches the main computation:
+        Bexternal = np.array(vcasing.compute_external_B(B1d))
+
+        # Unpack 1D array results:
+        Binternal1d = B1d - Bexternal
+        Binternal3d = np.zeros((nphi, ntheta, 3))
+        for jxyz in range(3):
+            Binternal3d[:, :, jxyz] = Binternal1d[jxyz * nphi * ntheta: (jxyz + 1) * nphi * ntheta].reshape((nphi, ntheta), order='C')
+
+        """
+        # Check order:
+        index = 0
+        for jxyz in range(3):
+            for jphi in range(nphi):
+                for jtheta in range(ntheta):
+                    np.testing.assert_allclose(Binternal1d[index], Binternal3d[jphi, jtheta, jxyz])
+                    index += 1
+        """
+
+        Btotal_normal = np.sum(B3d * unit_normal, axis=2)
+        Binternal_normal = np.sum(Binternal3d * unit_normal, axis=2)
+
+        vc = cls()
+        vc.ntheta = ntheta
+        vc.nphi = nphi
+        vc.theta = surf.quadpoints_theta
+        vc.phi = surf.quadpoints_phi
+        vc.B_total = B3d
+        vc.gamma = gamma
+        vc.unit_normal = unit_normal
+        vc.B_internal = Binternal3d
+        vc.B_internal_normal = Binternal_normal
+
+        if filename is not None:
+            if filename == 'auto':
+                directory, basefile = os.path.split(vmec.output_file)
+                filename = os.path.join(directory, 'vcasing' + basefile[4:])
+                logger.debug(f'New filename: {filename}')
+            vc.save(filename)
```

Comment:
I'd not call VirtualCasing.save automatically and instead ask the users to explicitly make a call to save the data.

Reply:
I'd prefer to keep saving as the default. The reason is that 99% of the time, the use case is to run virtual casing once, save the result, and load in the result many times for stage-2 coil optimizations with different parameters. If a user ever wants to not save a file, they can set `filename=None` when calling `VirtualCasing.from_vmec()`.